### PR TITLE
Middle Mouse Drag to Edit Values

### PR DESCRIPTION
**Pull request by turtletooth**
This is for #595 

This does a number of changes:

- All value fields can be middle click and dragged to change their value.
- All value fields in the tool option bar that don't have a slider can have their label left-clicked on and dragged to change the value.
- The function editor spreadsheet values can be ctrl + clicked and dragged to change their value.  (This also works on empty cells in columns that already have at least one key.  This is handy for quickly setting keys.)
- Alt + click on a key in the function editor will clear they key.

I was torn on using middle mouse for everything except the function editor.  I wanted the commands to be the same everywhere.  The benefit of middle mouse is that it doesn't require two hands and a tablet pen button can often be set to middle mouse - making it useful for tablet users.  However, middle mouse can't be used on the function editor, since it will pan the spreadsheet and that is useful functionality.  

Because of this I opted for different commands for the function editor and all other value fields- trying to do what is most convenient for users.  ( I realize that remembering two commands isn't convenient, but regularly using two hands for a command isn't also.)

I personally am most excited about the function editor commands as they seem to make adjusting key values and setting/clearing keys a breeze.

### CODE REVIEW

**Comment by blurymind** [comment]
Thank you! It should really help people with smaller screens. One
suggestion to add to it- mid click drag + shift could change the value at
smaller increments .. similar to how it is in blender

**Comment by artisteacher** [comment]
I'm really excited to have this functionality. I do wish it was left-click and drag like Blender and Krita. I almost always just use  a touchpad and am not sure it is possible to middle click and drag on a touchpad.

**Comment by blurymind** [comment]
We could use left click as well, it will work better with tablets too..
An alternative way to do it that would be neat for pc users is also mouse
scrolling :)

**Comment by artisteacher** [comment]
If left-click and drag isn't possible, scrolling would be a great solution. The addition of shift for more precision is a good idea.

**Comment by blurymind** [comment]
@turtletooth it doesnt have to use labels. Click dragging on input boxes actually makes a lot sense imo

**Comment by turtletooth** [comment]
Question about the function editor- I currently have it so you can ctrl+click and drag on a value in the function editor spreadsheet to change the value.  (This can be used to also set keyframes on unkeyed cells as long as the column has a key.)  I also set alt+click to remove keys.  

So far this makes editing values much faster than double clicking and typing the value.

The question is "What should the keys be that go with the mouse move?"
Shift + Click
Ctrl + Click
Alt + Click

There needs to be some modifier because click on its own needs to be available for selecting cells, and middle mouse is used to pan the spreadsheet.

**Comment by turtletooth** [comment]
I am also working on getting label click and drag working for only the edit tool and selection tool.  It's doable, but tedious.  But since I also use tablets alot, it would be a nice option to have.   Most other tool values have a slider. Plastic tool seems to be the only other one that doesn't - I'll look at that too.

All values will be able to do the middle mouse button as well.

**Comment by openanim** [comment]
First of all, thank you for your work, I don't understand everything, I don't know if it is related or not but i wanted to know 2 things

- We will always be able to navigate between the different drawings , for example to navigate frame by frame, skipping 3 frames by 3 frames, etc. thanks to the mouse wheel or ctrl+mouse wheel ?

2) If it will be possible to change the values ​​of the brush and geometry tools with this ? Or if I am completely off topic ?

**Comment by turtletooth** [comment]
Almost done- first post edited to include current functionality.

Todo:

- [ ] Adjust drag value if the measure in a field is inches or cm.
- [ ] Add better comments
- [ ] Squash commits

**Comment by turtletooth** [comment]
Done - man that was a lot more involved than I thought it would be.

**Comment by blurymind** [comment]
@turtletooth you are on fire! Thank you. It's perfect :D

**Comment by blurymind** [comment]
i think ctrl would be a good default modifier, but ideally it should be configurable.
Mid click drag is perfect imo

**Comment by RodneyBaker** [comment]
These are great time savers.
As I click and drag my middle mouse my my fingers and brain simultaneously say, "Smooooooth."  :)

**Comment by turtletooth** [comment]
Don't forget to try clicking the labels of the edit tool with left click, or ctrl + left click and drag on the function viewer.  Lots of fun.

**Comment by RodneyBaker** [comment]
@turtletooth said:
Don't forget to try clicking the labels of the edit tool with left click, or ctrl + left click and drag on the function viewer. Lots of fun.

I think I'm following all of that.
clicking
labels
labels of the edit tool
click with a left click (or ctrl click) and then drag
click and drag on function viewer
...
Okay, I think here is where I get lost.
The function viewer is the pink gnomon in the center of the canvas right?

Aside:  And... I never realized we could scale the Table itself.  Very interesting.

**Comment by shun-iwasawa** [comment]
Jenkins

File: toonz/sources/toonzqt/doublefield.cpp [hunk 1]
```diff
@@ -461,6 +494,58 @@ void MeasuredDoubleLineEdit::timerEvent(QTimerEvent *) {
...
+    m_mouseDragEditing = true;
+  } else
+    QLineEdit::mousePressEvent(e);
+}
+
+//-----------------------------------------------------------------------------
+
+void MeasuredDoubleLineEdit::mouseMoveEvent(QMouseEvent *e) {
+  if ((e->buttons() == Qt::MiddleButton) || m_labelClicked) {
+    m_value->modifyValue((e->x() - m_xMouse) / 2);
```
Comment (shun-iwasawa) [inline toonz/sources/toonzqt/doublefield.cpp]: In some Fx there are parameters with small value range. For instance, the `Opacity` parameter of the `Over Ino Fx` has 0.0-1.0 value range. In such case middle-dragging of the field cannot get the values other than 0 and 1 for now.

How about modifying this line, adding the `precision` argument like as follows ?:
```
 int precision = (m_maxValue > 100) ? 0 : ((m_maxValue > 10) ? 1 : 2);
 m_value->modifyValue((e->x() - m_xMouse) / 2, precision);
```
I once thought of referring `MeasuredDoubleLineEdit::m_decimals` variable but using the maximum value of the field seems to perform better.

File: toonz/sources/toonzqt/doublefield.cpp [hunk 2]
```diff
@@ -461,6 +494,58 @@ void MeasuredDoubleLineEdit::timerEvent(QTimerEvent *) {
...
+    QLineEdit::mouseMoveEvent(e);
+}
+
+//-----------------------------------------------------------------------------
+
+void MeasuredDoubleLineEdit::mouseReleaseEvent(QMouseEvent *e) {
+  if ((e->buttons() == Qt::NoButton && m_mouseDragEditing) || m_labelClicked) {
+    m_xMouse   = -1;
+    m_modified = true;
+    onEditingFinished();
```
Comment (shun-iwasawa) [inline toonz/sources/toonzqt/doublefield.cpp]: How about inserting `clearFocus();` just below this line?

**Comment by shun-iwasawa** [comment]
Related to this feature, I found a crash problem when dragging the frame field in the console and modified in #1574 .
Anyway this feature is very useful. Thanks @turtletooth !

**Comment by turtletooth** [comment]
Thanks for the review @shun-iwasawa - I made the requested changes.

**Comment by shun-iwasawa** [comment]
Jenkins

File: toonz/sources/include/toonzqt/lineedit.h [hunk 3]
```diff
@@ -33,6 +33,7 @@ class DVAPI LineEdit : public QLineEdit {
 
   bool m_isReturnPressed;
   bool m_forbiddenSpecialChars;
+  bool m_mouseDragEditing;
```
Comment (shun-iwasawa) [inline toonz/sources/include/toonzqt/lineedit.h]: @turtletooth Sorry for the late notice, but it seems that `bool m_mouseDragEditing` should be initialized to `false` here or in ctor.

**Comment by turtletooth** [comment]
Done.  Thanks @shun-iwasawa

**Comment by shun-iwasawa** [comment]
Jenkins again

**Comment by shun-iwasawa** [comment]
LGTM Thank you @turtletooth for addition !

**Comment by artisteacher** [comment]
I love this! It could also be helpful for text size. Just a thought, but replacing the sliders in the tool option bar with the click + drag on the label could save a lot of room in the tool option bar.

**Pull request merged**